### ISSUE TITLE
ErizoJS timed out will be removed

### ISSUE THREAD

**Issue by kekkokk** [lynckia/licode#969]
**Description**

It happens sometimes, that I get:

> message: ErizoJS timed out will be removed, erizoId: 4c556da2-b216-c20d-5344-50d1fc4c456a, publishersAffected: 1

**Steps to reproduce the issue:**
1. edit basicexample and make a function that unpublish and immediately after republish a stream
2. open it and try to call that function few times

**Describe the results you received:**

![screen shot 2017-07-14 at 12 50 06 pm](https://user-images.githubusercontent.com/15072723/28209527-28ea8028-6893-11e7-9e2b-d5428b66797f.png)

it works, but when I click "republish" again this often happens:

![screen shot 2017-07-14 at 12 47 30 pm](https://user-images.githubusercontent.com/15072723/28209443-d0c91e90-6892-11e7-807d-b1f1bf0fa5cd.png)

**Describe the results you expected:**

the structures in roomcontroller are cleared correctly

**Additional information you deem important (e.g. issue happens only occasionally):**

it happens especially when the addpublisher encounter a timeout and make a retry.

It's not a real problem, because the old publisher is gone, but maybe it's a sign that something isn't managed as expected. 
Also happened to me (when 4/5 people were in a many to many) that this function killed my js with all the persons still in it. (maybe this is an unconnected problem, maybe the js was full and didn't have enough cpu). anyway, investigating this problem I found this

The complete log:
[log.txt](https://github.com/lynckia/licode/files/1148164/log.txt)

**Licode commit/release where the issue is happening**

https://github.com/lynckia/licode/pull/965

**Additional environment details (Local, AWS, Docker, etc.):**

local end ec2

**Comment by kekkokk** [comment]
any news? I constantly get 

> [2017-07-25 07:53:16.557] [ERROR] RoomController - message: ErizoJS timed out will be removed, erizoId: 2a020f35-788c-2719-f598-3ba78f337c5d, publishersAffected: 1 <i-0b0b56af0022e9856>

in prod env and I really don't know how to solve the issue. this is very unreliable at the moment

**Comment by kekkokk** [comment]
ok so: a more detailed debug:
steps to reproduce:
open basicexample
publish a stream
unpublish that stream and immediately publish another stream
90% times this will cause last erizojs to close and a new one to start, BUT the new stream that we are going to publish goes on timeout 'cause the erizojs just started is not ready yet.
during this retry we close the browser.
after that, roomcontroller successfully add the stream to an erizojs but the publisher doesn't exist anymore.
result: one js stuck with a "ghost" publisher.

**Comment by arsenicraghav** [comment]
@kekkokk 
Have you tried looking at TIMEOUT_LIMIT in roomController.js which is set to 2 currently ? You may wanna increase or decrease it and see what happens.

**Comment by kekkokk** [comment]
Yeah but unfortunately there's a bad implementation of the mechanism.
If you think at what I wrote you can immediately comprehend where the problem is.
the retry shouldn't be done at server side but at client side.

**Comment by jcague** [comment]
what's your config values for the number of erizoJS processes?

**Comment by kekkokk** [comment]
to reproduce with 100% success i suggest you 1 process.
I fixed it by doing the retry logic client side. 
The problem is that i'm stuck at licode v3 with the new client may take some times to me to do a pr.

basically, if you put a console log inside the keepalive:
`
var sendKeepAlive = function() {
        for (var e in erizos) {
            console.log(">>> KA ERIZO: " + e + " pub: [" + erizos[e].publishers + "] count: " + erizos[e].kaCount);
            amqper.callRpc('ErizoJS_' + e, 'keepAlive', [], {callback: callbackFor(e)});
        }
    };
`

then make a function in basicexample that does:
`
function republish(){
room.unpublish(localstream);
room.publish(localstream);
}
`

you may be able to reproduce joining the basicexample, call that function and immediately close the tab.

you will see from the console log that there's a publisher stuck in the erizojs.

**Comment by jcague** [comment]
Yep, I'd say then that there's an issue while disconnecting and connecting the exacting same publisher. It should be fixed if you either create another localStream or wait a second or so between unpublish and publish, am I right?

**Comment by kekkokk** [comment]
no it happens with also other streams.

just think at this case:
I publish a stream, the js is not ready yet so we're stuck at the roomcontroller addpublisher.
it receive a timeout and then it recursively call addpublisher until the max retries are reached or it succeded.
but while he's waiting the 'timeout' coming from rabbit, the client disconnects.
the client disconnecting doesn't send any unpublish events because he hasn't received the callback from the controller assigning an id to that stream. (it receive the callback only when the addpublisher succeded and its in 'initializing' state, but remember, we're stuck at the erizo timeout right now.

so the client has disconnected and 1 second after (the js is ready now and the second retry from the controller succeeded) the roomcontroller try to send the callback with the stream id assigned to the client that doesn't exist anymore. 

that js will never close itself again because seems that there's one publisher inside (not true)

**Comment by zevarito** [comment]
guys, just landing here, could this be related: https://github.com/lynckia/licode/commit/baf793d70f012a9ffed5836406ebac8a2b9d119f ?

**Comment by kekkokk** [comment]
don't think so
